File: MPP-Java/src/main/java/repository/ReservationDBRepository.java
package repository;

import domain.Reservation;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.List;
import java.util.Properties;

public class ReservationDBRepository implements ReservationRepository{

    private JdbcUtils dbUtils;
    private static final Logger logger= LogManager.getLogger();

    public ReservationDBRepository(Properties props) {
        logger.info("Initializing ReservationDBRepository with properties: {} ",props);
        dbUtils=new JdbcUtils(props);
    }

    @Override
    public Reservation findOne(Long aLong) {
        logger.traceEntry();
        Connection con = dbUtils.getConnection();
        try(PreparedStatement preparedStatement = con.prepareStatement("select * from Reservations where id=?")){
            preparedStatement.setLong(1,aLong);
            try(ResultSet result = preparedStatement.executeQuery()){
                if(result.next()){
                    String name = result.getString("name");
                    Long phone = result.getLong("phone");
                    Long tickets = result.getLong("tickets");
                    Long excursionId = result.getLong("excursionId");
                    Reservation reservation = new Reservation(name, phone, tickets, excursionId);
                    logger.traceExit(reservation);
                    return reservation;
                }
            }
        }
        catch (SQLException ex){
            logger.error(ex);
            System.out.println(ex);
        }
        logger.traceExit("No reservation with id {} found", aLong);
        return null;
    }

    @Override
    public List<Reservation> findAll() {
        logger.traceEntry();
        Connection con = dbUtils.getConnection();
        List<Reservation> reservations = new ArrayList<>();
        try(PreparedStatement preparedStatement = con.prepareStatement("select * from Reservations")){
            try(ResultSet result = preparedStatement.executeQuery()) {
                while(result.next()) {
                    Long id = result.getLong("id");
                    String name = result.getString("name");
                    Long phone = result.getLong("phone");
                    Long tickets = result.getLong("tickets");
                    Long excursionId = result.getLong("excursionId");
                    Reservation reservation = new Reservation(name, phone, tickets, excursionId);
                    reservation.setId(id);
                    reservations.add(reservation);
                }
            }
        }
        catch (SQLException ex){
            logger.error(ex);
            System.err.println("Error DB"+ex);
        }
        logger.traceExit(reservations);
        return reservations;
    }

    @Override
    public void save(Reservation entity) {
        logger.traceEntry("Saving reservation {}", entity);
        Connection con = dbUtils.getConnection();
        try(PreparedStatement preparedStatement = con.prepareStatement("insert into Reservations(name, phone, tickets, excursionId) values (?,?,?,?)")){
            preparedStatement.setString(1, entity.getName());
            preparedStatement.setLong(2,entity.getPhoneNumber());
            preparedStatement.setLong(3, entity.getTicketsNumber());
            preparedStatement.setLong(4, entity.getExcursionID());
            int result = preparedStatement.executeUpdate();
            logger.trace("Saved {} instances", result);
        }
        catch (SQLException ex){
            logger.error(ex);
        }
        logger.traceExit();
    }


    @Override
    public void delete(Long aLong) {
        logger.traceEntry();
        Connection con = dbUtils.getConnection();
        try(PreparedStatement preStmt=con.prepareStatement("delete from Reservations where id=?")){
            preStmt.setLong(1,aLong);
            int result=preStmt.executeUpdate();
            logger.trace("Deleted {} instances", result);
        }catch (SQLException ex){
            logger.error(ex);
            System.out.println("Error DB "+ex);
        }
        logger.traceExit();
    }

    @Override
    public void update(Long aLong, Reservation entity) {
        logger.traceEntry("update reservation {}", entity);
        Connection con = dbUtils.getConnection();
        try(PreparedStatement preparedStatement = con.prepareStatement("update Reservations set name = ?, phone = ?, tickets = ?, excursionId = ? where id = ?;")){
            preparedStatement.setString(1, entity.getName());
            preparedStatement.setLong(2, entity.getPhoneNumber());
            preparedStatement.setLong(3, entity.getTicketsNumber());
            preparedStatement.setLong(4, entity.getExcursionID());
            preparedStatement.setLong(5,aLong);
            int result = preparedStatement.executeUpdate();
            logger.trace("Updated {} instances", result);
        }
        catch (SQLException ex){
            logger.error(ex);
        }
        logger.traceExit();
    }

}

File: CSharp/CSharp/repository/ReservationDBRepository.cs
using System;
using System.Collections.Generic;
using System.Data;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using CSharp.domain;
using CSharp.domain.validators;
using CSharp.Repository;
using log4net;

namespace CSharp.repository
{
    public class ReservationDBRepository : IReservationRepository
    {


        private static readonly ILog log = LogManager.GetLogger("ReservationDBRepository");
        private readonly IValidator<Reservation> reservationValidator;

        public ReservationDBRepository(IValidator<Reservation> validator)
        {
            log.Info("Creating ReservationDBRepository");
            this.reservationValidator = validator;
        }
        public IEnumerable<Reservation> FindAll()
        {
            IDbConnection con = JdbcManager.GetConnection();
            IList<Reservation> reservations = new List<Reservation>();

            using (var command = con.CreateCommand())
            {
                command.CommandText = "select id, name, phone, tickets, excursionId from Reservations";

                using (var dataReader = command.ExecuteReader())
                {
                    while (dataReader.Read())
                    {
                        long idRes = dataReader.GetInt64(0);
                        string name = dataReader.GetString(1);
                        long phone = dataReader.GetInt64(2);
                        long tickets = dataReader.GetInt64(3);
                        long excursionId = dataReader.GetInt64(4);

                        Reservation reservation = new Reservation(name, phone, tickets, excursionId);
                        reservation.Id = idRes;
                        reservations.Add(reservation);
                    }
                }
            }
            log.Info(reservations);
            return reservations;
        }

        public void Save(Reservation entity)
        {
            log.InfoFormat("Entered save with entity {0}", entity);
            IDbConnection con = JdbcManager.GetConnection();
            try
            {
                reservationValidator.validate(entity);
            }
            catch(ValidationException ex)
            {
                throw new RepositoryException(ex.Message);
            }
            using (var command = con.CreateCommand())
            {
                command.CommandText = "insert into Reservations(name, phone, tickets, excursionId) values (@name, @phone, @tickets, @excursionId)";
                var paramName = command.CreateParameter();
                paramName.ParameterName = "@name";
                paramName.Value = entity.Name;
                command.Parameters.Add(paramName);

                var paramPhone = command.CreateParameter();
                paramPhone.ParameterName = "@phone";
                paramPhone.Value = entity.PhoneNumber;
                command.Parameters.Add(paramPhone);

                var paramTickets = command.CreateParameter();
                paramTickets.ParameterName = "@tickets";
                paramTickets.Value = entity.TicketsNumber;
                command.Parameters.Add(paramTickets);

                var paramId = command.CreateParameter();
                paramId.ParameterName = "@excursionId";
                paramId.Value = entity.TicketsNumber;
                command.Parameters.Add(paramId);

                var result = command.ExecuteNonQuery();
                if (result == 0)
                    Console.WriteLine("No reservation added!");
            }
        }

        public Reservation FindOne(long id)
        {
            log.InfoFormat("Entering findOne with value {0}", id);

            IDbConnection con = JdbcManager.GetConnection();

            using (var command = con.CreateCommand())
            {
                command.CommandText = "select id, name, phone, tickets, excursionId from Reservations where id = @id";
                IDbDataParameter paramId = command.CreateParameter();
                paramId.ParameterName = "@id";
                paramId.Value = id;
                command.Parameters.Add(paramId);

                using (var dataReader = command.ExecuteReader())
                {
                    if (dataReader.Read())
                    {
                        long idRes = dataReader.GetInt64(0);
                        string name = dataReader.GetString(1);
                        long phone = dataReader.GetInt64(2);
                        long tickets = dataReader.GetInt64(3);
                        long excursionId = dataReader.GetInt64(4);


                        Reservation reservation = new Reservation(name, phone, tickets, excursionId);
                        reservation.Id = idRes;

                        log.InfoFormat("Exiting findOne with value {0}", reservation);
                        return reservation;
                    }
                }
            }
            log.InfoFormat("Exiting findOne with value {0}", null);

            return null;
        }

        public void Delete(long id)
        {
            log.InfoFormat("entering delete with id {0}", id);
            IDbConnection con = JdbcManager.GetConnection();
            using (var command = con.CreateCommand())
            {
                command.CommandText = "delete from Reservations where id = @id";
                var paramId = command.CreateParameter();
                paramId.ParameterName = "@id";
                paramId.Value = id;
                command.Parameters.Add(paramId);

                var dataR = command.ExecuteNonQuery();
                if (dataR == 0)
                {
                    Console.WriteLine("No reservation deleted!");
                }
            }
        }

        public void Update(long id, Reservation entity)
        {
            IDbConnection con = JdbcManager.GetConnection();
            using (var command = con.CreateCommand())
            {
                command.CommandText = "update Reservations set name = @name, phone = @phone, tickets = @tickets, excursionId = @excursionId where id = @id";
                var paramName = command.CreateParameter();
                paramName.ParameterName = "@name";
                paramName.Value = entity.Name;
                command.Parameters.Add(paramName);

                var paramPhone = command.CreateParameter();
                paramPhone.ParameterName = "@phone";
                paramPhone.Value = entity.PhoneNumber;
                command.Parameters.Add(paramPhone);

                var paramTickets = command.CreateParameter();
                paramTickets.ParameterName = "@tickets";
                paramTickets.Value = entity.TicketsNumber;
                command.Parameters.Add(paramTickets);

                var paramIdExcursion = command.CreateParameter();
                paramIdExcursion.ParameterName = "@excursionId";
                paramIdExcursion.Value = entity.TicketsNumber;
                command.Parameters.Add(paramIdExcursion);

                var paramId = command.CreateParameter();
                paramId.ParameterName = "@id";
                paramId.Value = id;

                var dataR = command.ExecuteNonQuery();
                if (dataR == 0)
                {
                    Console.WriteLine("No user updated!");
                }
            }
        }
    }
}